### PR TITLE
Added VectorTools::coordinate_space_integration

### PR DESCRIPTION
**Pull request by yl3dy**

### CODE REVIEW

File: include/hyper.deal/numerics/vector_tools.h [hunk 1]
```diff
@@ -324,6 +324,92 @@ namespace hyperdeal
...
+      dealii::
+        FEEvaluation<dim_v, degree, n_points, 1, Number, VectorizedArrayTypeV>
+         phi_v(data.get_matrix_free_v(), dof_no_v, 0/*dummy*/);
+
+      // Clear destination vector
+      dst = 0.0;
+
+      constexpr auto N_v_points = dealii::Utilities::pow(n_points, dim_v);
+
+      VectorizedArrayTypeX scratch[N_v_points];
```
Comment (peterrum) [inline include/hyper.deal/numerics/vector_tools.h:364]:
```suggestion
      AlignedVector<VectorizedArrayTypeX> scratch(N_v_points);
```

File: include/hyper.deal/numerics/vector_tools.h [hunk 2]
```diff
@@ -324,6 +324,92 @@ namespace hyperdeal
...
+
+      data.template cell_loop<Vector_Out, Vector_In>(
+        [&](const auto&,
+            Vector_Out&      dest,
+            const Vector_In& src_,
```
Comment (peterrum) [inline include/hyper.deal/numerics/vector_tools.h:369]:
```suggestion
            Vector_Out&      dst,
            const Vector_In& src,
```

File: include/hyper.deal/numerics/vector_tools.h [hunk 3]
```diff
@@ -324,6 +324,92 @@ namespace hyperdeal
...
+          const auto N_lanes_x = phi.n_vectorization_lanes_filled();
+
+          // SIMD-powered summing over all x points
+          for (unsigned int qv = 0; qv < phi_v.n_q_points; ++qv) {
+            scratch[qv] = 0.0;
+            for (unsigned int qx = 0; qx < phi_x.n_q_points; ++qx)
+              scratch[qv] += data_ptr_src[qx + qv*N_v_points] * phi_x.JxW(qx);
+          }
+
+          // Accumulation over SIMD lanes
```
Comment (peterrum) [inline include/hyper.deal/numerics/vector_tools.h:389]:
```suggestion
          // perform cross-lane reduction
```

File: include/hyper.deal/numerics/vector_tools.h [hunk 4]
```diff
@@ -324,6 +324,92 @@ namespace hyperdeal
...
+          phi.read_dof_values(src_);
+          phi_x.reinit(cell.x);
+          phi_v.reinit(cell.v);
+
+          const VectorizedArrayType* data_ptr_src = phi.get_data_ptr();
+          VectorizedArrayTypeV* data_ptr_dst = phi_v.begin_dof_values();
+
+          const auto N_lanes_x = phi.n_vectorization_lanes_filled();
+
+          // SIMD-powered summing over all x points
```
Comment (peterrum) [inline include/hyper.deal/numerics/vector_tools.h:382]:
```suggestion
          // reduce in x-direction in a vectorized fashion
```

File: include/hyper.deal/numerics/vector_tools.h [hunk 5]
```diff
@@ -324,6 +324,92 @@ namespace hyperdeal
...
+            scratch[qv] = 0.0;
+            for (unsigned int qx = 0; qx < phi_x.n_q_points; ++qx)
+              scratch[qv] += data_ptr_src[qx + qv*N_v_points] * phi_x.JxW(qx);
+          }
+
+          // Accumulation over SIMD lanes
+          for (unsigned int qv = 0; qv < phi_v.n_q_points; ++qv) {
+            data_ptr_dst[qv] = 0.0;
+            for (unsigned int lane = 0; lane < N_lanes_x; ++lane)
+              data_ptr_dst[qv] += scratch[qv][lane];
```
Comment (peterrum) [inline include/hyper.deal/numerics/vector_tools.h:393]: I think there should be somewhere an assert that this only works for `VectorizedArrayTypeV::size()==1`.

File: include/hyper.deal/numerics/vector_tools.h [hunk 6]
```diff
@@ -324,6 +324,92 @@ namespace hyperdeal
                     comm);
     }
 
+    /** Perform coordinate space integration */
```
Comment (peterrum) [inline include/hyper.deal/numerics/vector_tools.h:327]: Could you adjust the comment so that it matches with the other ones?

**Review by peterrum — COMMENTED** [review]
@yl3dy Thanks for the PR! Looks very good already. A few request from my side:
- could you add a change-log entry in https://github.com/hyperdeal/hyperdeal/tree/master/doc/news/changes?
- I think you have not formatted the code! Do you know how to do that?

My suggestion would be to merge the PR soon (without a test) and I will rebase simply https://github.com/hyperdeal/hyperdeal/pull/92 so that we have the test for both functions in one go! Is that fine with you?

**Comment by yl3dy** [comment]
> My suggestion would be to merge the PR soon (without a test) and I will rebase simply #92 so that we have the test for both functions in one go! Is that fine with you?

I initially planned to add patches with tests manually, but your suggestion would be better.

**Comment by peterrum** [comment]
We are still using clang-format 6 (as did deal.II not that long ago). I am planing to update its version (see https://github.com/hyperdeal/hyperdeal/issues/85) but am a bit reluctant, since I don't want to introduce too many merge conflicts in @kkormann's code. You can get the old version of clang-format with the following script: https://github.com/hyperdeal/hyperdeal/blob/master/contrib/utilities/download_clang_format

**Review by peterrum — APPROVED** [review]
Thanks a lot @yl3dy! Looking forward to future additions!

I'll merge once the tests have passed!

**Pull request merged**